Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -1038,14 +1034,20 @@ extension TaskHandler: ChannelDuplexHandler {
                 self.failTaskAndNotifyDelegate(error: error, self.delegate.didReceiveError)
             }
         default:
-            self.state = .endOrError
-            self.failTaskAndNotifyDelegate(error: error, self.delegate.didReceiveError)
+            switch self.state {
+            case .idle, .bodySent, .sent, .head, .redirected, .body:
+                self.state = .endOrError
+                self.failTaskAndNotifyDelegate(error: error, self.delegate.didReceiveError)
+            case .endOrError:
+                // error was already handled
+                break
+            }
         }
     }
 
     func handlerAdded(context: ChannelHandlerContext) {
         guard context.channel.isActive else {
-            self.failTaskAndNotifyDelegate(error: HTTPClientError.remoteConnectionClosed, self.delegate.didReceiveError)
+            self.errorCaught(context: context, error: HTTPClientError.remoteConnectionClosed)
```

Comment:
There are a bunch of changes in this file that don't seem well motivated: what are they there for?

Reply:
These were cause afer the base branch was changed from `master` to `main`, sorry about that, I'll clean it up.

Reply:
I'm changing the base branch back from `main` to `master`, because `main` is outdated compared to `master` in this repo.

Reply:
@ktoso What happened with the migration to `main` here?

Reply:
Something's wrong... I have it noted that we completed the work here but seems master still was the main thing, wat?

Please do not target `master` @MaxDesiatov, please keep targeting main.
I'll bring main up to date again and re-do the steps about branch protection.
CI is also ticked off to have been changed but I can't confirm that since no admin rights there.

Reply:
It seems we/I missed to change the default branch so a few PRs were made against master and merged. I've brought main up to date again, fixed that, re targeted this at main and _removed_ master.

Please keep using main from now-on, it should be exactly the same as master just was (it was behind a few commits indeed).

Reply:
Thanks, I've rebased it on top of `main`. @Lukasa now ready for review.